squad/core/management/commands/compute_project_statuses.py

```diff
@@ -0,0 +1,88 @@
+from django.utils import timezone
+
+from squad.core.models import Project, Build, ProjectStatus
+
+
+logger = logging.getLogger()
+environments_cache = {}
+
+
+def valid_date(date):
+    try:
+        return datetime.strptime(date, "%Y-%m-%d")
+    except ValueError:
+        msg = "'{0}' is not a valid date.".format(
+            date)
+        raise argparse.ArgumentTypeError(msg)
+
+
+class Command(BaseCommand):
+
+    help = """Compute metric summaries per build per environment"""
+
+    def add_arguments(self, parser):
+        parser.add_argument('--project', help='Optionally, specify a project to compute, on the form $group/$project')
+        parser.add_argument('--show-progress', action='store_true', help='Prints out one dot per build in stdout')
+        parser.add_argument(
+            '--start-date',
+            dest="start_date",
+            default=(datetime.now() - timedelta(days=180)),
+            type=valid_date,
+            help="Start date for project status updates (default: 6 months before current date, format: YYYY-MM-DD)."
+        )
+        parser.add_argument(
+            '--end-date',
+            dest="end_date",
+            default=datetime.now(),
+            type=valid_date,
+            help="End date for project status updates (default: today, format: YYYY-MM-DD)."
+        )
+
+    def __progress__(self, show):
+        if show:
+            self.stdout.write(".", ending="")
+            self.stdout._out.flush()
+
+    def handle(self, *args, **options):
+        start_date = timezone.make_aware(options['start_date'])
+        end_date = timezone.make_aware(options['end_date'])
+        project_name = options['project'] or False
+        show_progress = options['show_progress']
+        logger.info("Filtering builds from %s to %s" % (start_date, end_date))
+        builds = Build.objects.filter(
+            datetime__range=(start_date, end_date)
+        )
+
+        project = None
+        if project_name:
+            slugs = project_name.split('/')
+            if len(slugs) != 2:
+                logger.error('Project "%s" is malformed (should be group_slug/project_slug). Exiting...' % (project_name))
+                return
+
+            try:
+                group_slug, project_slug = slugs
+                project = Project.objects.get(group__slug=group_slug, slug=project_slug)
+            except Project.DoesNotExist:
+                logger.error('Project "%s" does not exist. Exiting...' % (project_name))
+                return
+
+            logger.info('Filtering builds from project "%s"' % (project_name))
+            builds = builds.filter(project=project)
+
+        logger.info('Computing metrics summary for %d builds' % (builds.count()))
+        if show_progress:
+            logger.info('Showing progress, one dot means one processed build')
+
+        for build in builds.all():
+            self.__progress__(show_progress)
+            ProjectStatus.create_or_update(build)
```

Comment:
I foresee 1% of buggy results with this approach. Example:

b1 (finished=True)
b2 (finished=False, previous_build=b1)
b3 (finished=False, previous_build=b1)

Let's say `b3` received all TestJobs and it got finished before `b2`. ProjectStatus will be stored with regressions/fixes calculated against `b1`. After sometime, `b2` will also have finished=True. So when running this command, `b3.baseline=None`, therefore `previous_build=b2` and ProjectStatus will be recalculated against `b2` instead of `b1`.

I'm not sure of the % of builds that fall in this scenario, but it's nice to be aware of. I guess there's no way of preventing that...

Reply:
IMHO this is OK. Since we'll have direct link to the baseline it will be clear how the regressions were calculated so I don't consider this a bug.

Reply:
Ok